### ISSUE TITLE
 Cannot destructure property 'manager' of '.for' as it is undefined.

### ISSUE THREAD

**Issue by amk221** [emberjs/ember.js#20708]
### 🔬 Minimal Reproduction
Minimal reproduction of an error in ember-modifier

```
 Cannot destructure property 'manager' of '.for' as it is undefined.
```

https://github.com/amk221/-ember-modifier-manager/commit/43ddd72fb1b781066d7c38ee4cfd39c1d07ba0b0#diff-91d02793f5c4fd5e02a401c33509cafcb6fc7613aef16d97afe18f4b4a080d5cR15

### 🌍 Environment

- Ember: - 5.8
- Ember-CLI: -
- Node.js/npm: -
- OS: -
- Browser: -

**Comment by NullVoxPopuli** [comment]
Does the issue appear in
- ember-source 5.9?
- gjs?

Thanks!

**Comment by amk221** [comment]
ember-source 5.9 = yes, still occurs

gjs = yes, it still occurs in our app, which is gjs. But, in terms of focusing on the minimal reproduction, no, it instead outputs 

```
Attempted to set the same type of manager multiple times on a value. You can only associate one manager of each type with a given value. Value was (unknown function)
```

**Comment by NullVoxPopuli** [comment]
That socond error seems like the deps got messed up. Does using pnpm reproduce nhe original error?

**Comment by amk221** [comment]
relooking

**Comment by amk221** [comment]
You're right.

Here's a gjs branch, with the original error

https://github.com/amk221/-ember-modifier-manager/commit/8443a5babec003e515a117ebd3dda58d963eca4a#diff-8b01f9fcdd7fd5370da514614ab30fe4ea2ebfd452dac4b7e4fd3482dc0e5421

**Comment by NullVoxPopuli** [comment]
In the gjs branch, there was an issue.
the modifier import was shadowing the keyword.

After the fix (https://github.com/amk221/-ember-modifier-manager/pull/1), the new error is 
```
Uncaught (in promise) TypeError: scheduledUpdateModifiers[Symbol.iterator]().next().value is undefined
    commit runtime.js:3904
    commit runtime.js:3972
    inTransaction runtime.js:3991
    Ember 3
    invoke backburner.js.js:272
    flush backburner.js.js:188
    flush backburner.js.js:344
    _end backburner.js.js:773
    _boundAutorunEnd backburner.js.js:509
    promise callback*buildNext/< backburner.js.js:26
    flush Ember
```

which is another issue :thinking:

**Comment by NullVoxPopuli** [comment]
It's possible that one of these caused the issue:
- https://github.com/glimmerjs/glimmer-vm/pull/1565
- https://github.com/glimmerjs/glimmer-vm/pull/1573

**Comment by NullVoxPopuli** [comment]
ah, no the repro has a problem

**Comment by NullVoxPopuli** [comment]
Here is the fix:
- don't have two click listeners changing data for the modifier at the same time

```diff
diff --git a/app/components/foo.gjs b/app/components/foo.gjs
index 2dd2640..56f7ba3 100644
--- a/app/components/foo.gjs
+++ b/app/components/foo.gjs
@@ -23,12 +23,7 @@ export default class Foo extends Component {
   close() {
     console.log("close");
     this.isOpen = false;
-  }
-
-  @action
-  setValue(value) {
-    console.log("setValue");
-    this.value = value;
+    this.value = "foo";
   }
 
   <template>
@@ -44,7 +39,6 @@ export default class Foo extends Component {
       type="button"
       class="close"
       {{on "click" this.close}}
-      {{on "click" (fn this.setValue "foo")}}
     >
       Close
     </button>
```

Now... _should_ it work? probably.
why doesn't it work? idk.

maybe @chancancode can provide some insights there.

**Comment by amk221** [comment]
Thanks for looking. Don't forget this is a minimal reproduction... in our actual app it's not two click handlers, it's a tracked property changing and causing a re-render, I think. 

Anyway, I really would like it to work, since the code is 'valid'

**Comment by NullVoxPopuli** [comment]
> Since, you can combine the click handler and the error still happens.

that's not what happens in my PR to your repro :thinking: 

> Anyway, I really would like this to work of course

yeah, the fact it doesn't work does seem bad

**Comment by betocantu93** [comment]
Im also having this error with ember-focus-trap

> Attempted

**Comment by amk221** [comment]
After having ran ember-cli-update from 5.9.0 to 6.1.0 this is occurring a lot more in our app now

```
Uncaught (in promise) Error: Attempted to set the same type of manager multiple times on a value. You can only associate one manager of each type with a given value. Value was (unknown function)
```

As already mentioned, it may be an addon author applying  a modifier, and then, a consumer of the addon applying one of the same type again. Which should be allowed because they aren't in control of 'the fix'.

As you have already seen my minimal re-production repo, [here](https://github.com/zestia/ember-select-box/blob/72b4d54653327b992b1c7f3b62065350ca83e082/tests/integration/select-box/index/edge-case-test.gjs#L12) is a more realistic scenario.

**Comment by amk221** [comment]
Fixed by avoiding name clash

```gjs
import { modifier as _modifier } from 'ember-modifier';
const mod1 = _modifier(() => {})

//...

<template><div {{modifier mod2 ...}}></div></template>
```

**Comment by amk221** [comment]
New replication but a slightly different scenario. Please re-open


https://github.com/amk221/-ember-modifier-manager-issue/commit/8d6850b9cdd5918d1885bfaf344406b8564bac05

**Comment by NullVoxPopuli** [comment]
[In the REPL](https://limber.glimdown.com/edit?c=JYWwDg9gTgLgBAYQuCA7Apq%2BAzKy4DkAAgOYA2oI6UA9AMbKQZYEDcAUKJLHAN5wwoAQzoBrdABM4AXzi58xcpWo1BI0cFQk2nFDwCyECcGzBqcvCELoQAI2oBaEEZNmoOrtHj8o6EsABnGGoAEXQgqABXOhhoGQsFIht7WglwwQgATyFbMnQPPW84NHj5K2JklWdjU2oddjoyIQCAuBBMwxq3OHQAD2DUCVbO13Nedjg4BlQI6NioAAoASj4JybgAyLBqBYA6faEoEkiqLACljnW4X38g0PSomOgFmAALQIAaAXeA3YB9YKHEIQADuqAua2k7DW1RMmQW6DypxgX0gAWAMGAaCEZC%2BqCEVAkK3GVzegV2h2EmTgAF44PjCRSoFTLutpgEIHldmQICQFgRYdhMgQvmTfpShJkIVcTHAFgBCMW7YwSADK6BgW2Ja1JP3%2BAQ1AFUwMtWZMoeboZM-gaYMblqsruzOehubz%2BbatgRpeslSr1ZqwLSBFF0KyoWsAeggaDUMGHTSAHyOtloDlcnl8giAqASWPe8PsCPsPrceBpbBCSJkeCNZqtMLOHr9TBDRCMNCYbxrIhqMSSHr43Jhnt98RSCXUukAbQIQhFhFsC4IdAIAF0OGtYiRyOh4yskynJs6M%2B7s7zdwWdd9yZgcnlg4q9Xfh%2BHN5M86rkBr3lp97TkxJVMZhdN0szzDZvzJLQr11clJ2Dad9l2JVJy%2BAgJHXN8rTgAAeYJwCaYJE2vXDjAANz4XgFllJIhwfBZBW6doRlqKA4EnGlUOZSUliWaRpBIq5Jl4XgAGJozoV4b3FHjqWaOAAB8MRsRSBOvK5RJUkABI%2BDSqJoSTXnUq5cJoCihPWXDbHYmhE2s2zLMmazIhgWI4zrFoaQAIiEbyBEybYfNsVz3P80SSm8xpgDEfylW3XcBKc9YABULwfJjqFImgQrctBLNI3L3KmJovO82w4sC9BgtCtBwt4SLotimTlQgL8qGgkgkv04E4Han9NBIbKivytYzIIsAiPQEjpCAA&format=gjs) -- how do I see the issue?

**Comment by amk221** [comment]
Run the tests

<img width="1279" alt="Image" src="https://github.com/user-attachments/assets/edf50d12-bde7-4c30-a645-e215e8a9c32a" />

**Comment by NullVoxPopuli** [comment]
[REPL with clicks](https://limber.glimdown.com/edit?c=JYWwDg9gTgLgBAYQuCA7Apq%2BAzKy4DkAAgOYA2oI6UA9AMbKQZYEDcAUKJLHAN5wwoAQzoBrdABM4AXzi58xcpWo1BI0cFQk2nFDwCyECcGzBqcvCELoQAI2oBaEEZNmoOrtHj8o6EsABnGGoAEXQgqABXOhhoGQsFIht7WglwwQgATyFbMnQPPW84X1Q0qABldBgYPKlZeStiZJUSsuoC7iK0eIbCJLsVZ2NTdo52OjIhAIC4EEzDYbc4dAAPYNKZhddzXnY4OAZUCOjYqAAKAEo%2BPf24AMiwajOAOlehKBJIqiwAi45b4p%2BQLBKBhY4xaBnGAAC0CABoBLCAs8APrBd4hCAAd1Qfxu0nYNyGJkyZ3QeW%2BMARkACwBgwDQQjICNQQioEiuuwBMMCz3ewkycAAvHBWey%2BVABf9bocAhA8s8yBASGcCMTsJkCAiecj%2BUJMniASY4GcAIQ657GCSVGAPTk3blI1EBKoAVTAl2l%2BwJ3sJ%2BxRLpg7su1wBsvl6EVytVgYeBENtwtVptD2FAii6GlBJuaPQGOxqDTIaFAD5QzK0HKFUqVQR0VAJAX41n2Nn2KtOnA0tghJEyPAJlMZmFnMs1pgJDMkChmN4bkQ1GJJMtWblM-PF%2BIpHrBSKANoEIRawi2Y8EOgEAC6Y32sRI5HQRaupfL%2B3D1ejdeVD%2BbDsRvMwHI8jTc0nUAtcsxvLsIHKZAqlhLQn2FMsuQrI4IyjWtGzuOCeS0X9HV5Hc0z3V5ngtHcEQICQr0gv1iiEMBgAkAAVb9gJFKZMlQOgTXQAA3TAYGfFC-zyeAgneeARQEoTnjoSJJSElj3hIKp5KVF0glVYx%2BIIgFxLgIQ00k2BngAR0iahMkqPIIXOAg%2BX0-ZDNsEyYCkiyrKgGzyXQezVWeU88T-IQNOAMRPVCrEhDpQFSmoG0akkKKAVscLIoTfYhBiuLWkSqpkokVLbjCiYItEErsty%2BB8oqQraiquB0vKzKvSMmr4rKJLGqyuB8XogAeYJwEmYISz-QbdL4XgzmNJJV2As51SWOYthGKAjMlfUhQo7aDQuaRpAmgF9l4XgAGI8zoaF-11fajJmAAfOkbCeo6-wBc7XpAI64U%2BmaaGu6EPoBQaaF0k7bkG2xNpoEsYbhqH9hhyJqm6QdpiFAAiIRsYETJHhx2w0diVB8fO7psda0R8YtO8HyO5HbjY%2B9gJW6hJpoEn0dQKHJp5smDkmLHsdsOnCfQYnSbQCneCpmm6adRtYKoPCSCZgHMTgVX4M0EgucFtAJsm6EoBNsGjcLSnC3PCgxAIO7nmEJjWPY9AmaiVBwatqHwZGsAxvQCbpCAA&format=gjs) 

gonna try your 6.3 app now (the REPL is using 6.8)

**Comment by NullVoxPopuli** [comment]
I think this has something to do with the route template and controller combo. if I just use components, the problem goes away.

**Comment by amk221** [comment]
I ran ember-cli-update ([commit](https://github.com/amk221/-ember-modifier-manager-issue/commit/2ca7c9897ca8b5c43878766d88ce4408b8b002f3#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L59)), but the error is still there

**Comment by NullVoxPopuli** [comment]
If you're wanting to be unblocked -- can you try moving to components instead of controller + template? does that fix your issue?

**Comment by amk221** [comment]
Thanks its a good idea, but I [tried it](https://github.com/amk221/-ember-modifier-manager-issue/commits/main/): First by removing use of controllers, and Secondly by switching to ember-route-templates to make doubly sure.

**Comment by NullVoxPopuli** [comment]
did I mess something up here? test passes: https://github.com/amk221/-ember-modifier-manager-issue/pull/1/files#r2337741343

**Comment by amk221** [comment]
Well its not a true representation of the minimal reproduction of the issue, its not using arguments

**Comment by NullVoxPopuli** [comment]
[Here is a repro in the REPL](https://limber.glimdown.com/edit?c=JYWwDg9gTgLgBAYQuCA7Apq%2BAzKy4DkAAgOYA2oI6UA9AMbKQZYEDcAUKJLHAN5wwoAQzoBrdABM4AXzi58xcpWo1BI0cFQk2nFDwCyECcGzBqcvCELoQAI2oBaEEZNmoOrtHj8o6EsABnGGoAEXQgqABXOhhoGQsFIht7WglwwQgATyFbMnQPPW84X1Q0qABldBgYPKlZeStiZJUSsuoC7iK0eIbCJLsVZ2NTdo52OjIhAIC4EEzDYbc4dAAPYNKZhddzXnY4OAZUCOjYqAAKAEo%2BPf24AMiwajOAOlehKBJIqiwAi45b4p%2BQLBKBhY4xaBnGAAC0CABoBLCAs8APrBd4hCAAd1Qfxu0nYNyGJkyZ3QeW%2BMARkACwBgwDQQjICNQQioEiuuwBMMCz3ewkycAAvHBWey%2BVABf9bocAhA8s8yBASGcCMTsJkCAiecj%2BUJMniASY4GcAIQ657GCSVGAPTk3blI1EBKoAVTAl2l%2BwJ3sJ%2BxRLpg7su1wBsvl6EVytVgYeBENtwtVptD2FAii6GlBJuaPQGOxqDTIaFAD5QzK0HKFUqVQR0VAJAX41n2NnZfAELCyFIRQAeYLgSbBEs3XvGABufF4Z2NSVZuXQJvVSzmWxGUDgesyQqIW4uF2k0hHAN4vAAxHm6NC4LvJfrNzMAD50myPw8O26nl8gQ9wj%2BnmhL2hd99l7GgJ2POBe1sDcaBLaDYMg6DImqboJimAIhQAIiELCBEyR5sNsFDYlQPDT26LCJmAMQ8KIWISHIdBD0g-YABVlSY2YXHXUcaGI1DUBHUcBNIg5JmmIi8JgAj0CIki0HI3hKOo2ib0bcpkCqWEtBYj9MTgTSqB5LQ%2BNEtARzAgcwCHdASzGdDpkMmAhGCZY1kwCQZiQFBmG8PZ6OEMRJGWec8mlQL1BCrc0wAbQIIQtUIWwkoIOgCAAXTGfYGK4kVizLLkKyOCMo1rXK8mbD8LUwHI8jTc0nVqhcs2yuANK0kySCLK5S3LfZw2raMCEbO5Op07QExyp0YpFWLXmeC0twREbMtav1iiEMBgAkDjGPqkUpkyVA6BNdBx0wGBesKj88ngIJ3ngEVzsu546EiSVLrY94SCqN6lRdIJVQnKqATgO7NzTB7YGeABHSJqEySo8ghc4CD5UHbgh2woZcmH4cR5H0FR1VnhSvEPyEf6aNET1NyxIQ6UBUpqBtGpJE9D9bGpsQ6aEBmmdaVmqnZiROYBKnVNpv56cZ%2BAhYqEXanF25ualvmBflzzheqZWE2zUDrNs1ioM7YBuzgD99jnOq5NPGqwuY2QrZvCq7d4C03ZAsGiA64yJqFe2nT97TNBIb2AVvAVA49ma70yCO4Dgj9e2hKATeQwSp26NKKDEAhEV5YRtt2zi8hYqJUDA8yhL4o3XLs1tCX7GwbIbpDyhctzk6s1vjaAA&format=gjs)

**Comment by amk221** [comment]
<img width="751"  alt="Image" src="https://github.com/user-attachments/assets/44d03898-9b62-4ecd-a687-36c716eb8d41" />

**Comment by NullVoxPopuli** [comment]
yup -- and with the debug tools open, it can be debugged the same as your repro repo 🎉 

though, the error is slightly different. I was never able to see `.for` in the error

**Comment by amk221** [comment]
(Thats why I made the github repo, to be as close to the real problem in our app as possible!)

**Comment by amk221** [comment]
Please can the issue be reopened?

**Comment by NullVoxPopuli** [comment]
Based on the discussion and prior triage at the top of this issue, _i'd_ prefer a new issue so we start with a clean slate -- in this issue, there are like 5 different problems being looked at

e.g.:
<img width="1003" height="722" alt="Image" src="https://github.com/user-attachments/assets/86e0854d-817e-43ce-8b59-7e540dfca23b" />

good news though -- you already have a good reproduction, and that's a great place to start. 🎉
